solidity/contracts/AbstractECDSAKeep.sol

```diff
@@ -0,0 +1,559 @@
+                dividend
+            );
+        }
+
+        // Transfer of dividend for the last member. Remainder might be equal to
+        // zero in case of even distribution or some small number.
+        uint256 remainder = _value.mod(memberCount);
+        token.safeTransferFrom(
+            msg.sender,
+            beneficiaryOf(members[memberCount - 1]),
+            dividend.add(remainder)
+        );
+
+        emit ERC20RewardDistributed(_tokenAddress, _value);
+    }
+
+    /// @notice Gets current amount of ETH hold in the keep for the member.
+    /// @param _member Keep member address.
+    /// @return Current balance in wei.
+    function getMemberETHBalance(address _member)
+        external
+        view
+        returns (uint256)
+    {
+        return memberETHBalances[_member];
+    }
+
+    /// @notice Withdraws amount of ether hold in the keep for the member.
+    /// The value is sent to the beneficiary of the specific member.
+    /// @param _member Keep member address.
+    function withdraw(address _member) external {
+        uint256 value = memberETHBalances[_member];
+
+        require(value > 0, "No funds to withdraw");
+
+        memberETHBalances[_member] = 0;
+
+        /* solium-disable-next-line security/no-call-value */
+        (bool success, ) = beneficiaryOf(_member).call.value(value)("");
+
+        require(success, "Transfer failed");
+    }
+
+    /// @notice Submits a fraud proof for a valid signature from this keep that was
+    /// not first approved via a call to sign. If fraud is detected it tries to
+    /// slash members' KEEP tokens. For each keep member tries slashing amount
+    /// equal to the member stake set by the factory when keep was created.
+    /// @dev The function expects the signed digest to be calculated as a sha256
+    /// hash of the preimage: `sha256(_preimage))`. The function reverts if the
+    /// signature is not fraudulent. The function does not revert if KEEP slashing
+    /// failed but emits an event instead. In practice, KEEP slashing should
+    /// never fail.
+    /// @param _v Signature's header byte: `27 + recoveryID`.
+    /// @param _r R part of ECDSA signature.
+    /// @param _s S part of ECDSA signature.
+    /// @param _signedDigest Digest for the provided signature. Result of hashing
+    /// the preimage with sha256.
+    /// @param _preimage Preimage of the hashed message.
+    /// @return True if fraud, error otherwise.
+    function submitSignatureFraud(
+        uint8 _v,
+        bytes32 _r,
+        bytes32 _s,
+        bytes32 _signedDigest,
+        bytes calldata _preimage
+    ) external onlyWhenActive returns (bool _isFraud) {
+        bool isFraud = checkSignatureFraud(
+            _v,
+            _r,
+            _s,
+            _signedDigest,
+            _preimage
+        );
+
+        require(isFraud, "Signature is not fraudulent");
+
+        if (!fraudulentPreimages[_preimage]) {
+            punishMembersForSignatureFraud();
```

Comment:
I'd name it `slashForSignatureFraud`. We can apply any slashing strategy we want slash stake/seize bond, do both.